libraries/botbuilder/src/botFrameworkAdapter.ts

```diff
@@ -968,25 +979,87 @@ export class BotFrameworkAdapter extends BotAdapter implements IUserTokenProvide
      * Override this in a derived class to create a mock connector client for unit testing.
      */
     public createConnectorClient(serviceUrl: string): ConnectorClient {
-        if (BotFrameworkAdapter.isStreamingServiceUrl(serviceUrl)) {
+        return this.createConnectorClientInternal(serviceUrl, this.credentials);
+    }
+
+    /**
+     * Create a ConnectorClient with a ClaimsIdentity.
+     * @remarks
+     * If the ClaimsIdentity contains the claims for a Skills request, create a ConnectorClient for use with Skills.
+     * @param serviceUrl 
+     * @param identity ClaimsIdentity
+     */
+    public async createConnectorClientWithIdentity(serviceUrl: string, identity: ClaimsIdentity): Promise<ConnectorClient> {
+        if (!identity) {
+            throw new Error('BotFrameworkAdapter.createConnectorClientWithScope(): invalid identity parameter.');
+        }
+
+        const botAppId = identity.getClaimValue(AuthenticationConstants.AudienceClaim) ||
+            identity.getClaimValue(AuthenticationConstants.AppIdClaim);
+
+        // Anonymous claims and non-skill claims should fall through without modifying the scope.
+        let credentials: AppCredentials = this.credentials;
+
+        // If the request is for skills, we need to create an AppCredentials instance with
+        // the correct scope for communication between the caller and the skill.
+        if (botAppId && SkillValidation.isSkillClaim(identity.claims)) {
+            const scope = JwtTokenValidation.getAppIdFromClaims(identity.claims);
+            if (this.credentials.oAuthScope === scope) {
+                // Do nothing, the current credentials and its scope are valid for the skill.
+                // i.e. the adatper instance is pre-configured to talk with one skill.
+            } else {
+                // Since the scope is different, we will create a new instance of the AppCredentials
+                // so this.credentials.oAuthScope isn't overridden.
+                credentials = await this.getAppCredentials(botAppId, scope);
+
+                if (JwtTokenValidation.isGovernment(this.settings.channelService)) {
```

Comment:
In R8 we should discuss Skills in gov, and if yes test Skills in gov.

Reply:
The equivalent of this code is in [C#](https://github.com/microsoft/botbuilder-dotnet/blob/master/libraries/Microsoft.Bot.Builder/BotFrameworkAdapter.cs#L1168) and [Python](https://github.com/microsoft/botbuilder-python/blob/master/libraries/botbuilder-core/botbuilder/core/bot_framework_adapter.py#L648)... Are you saying we should remove it from all three languages? (I'm not opposed to this, as we haven't tested Skills in Gov)

Technically speaking, the type of Cloud shouldn't matter for parent-skill pairings as long as they're in the SAME cloud. Inter-cloud skills on the other-hand, it's not clear if there is any support for this.

There is _some_ interest in 3rd party channels on Gov Cloud, so we should factor code in a way that accounts for this possibly being a required scenario in the event that these 3rd party channels are NOT supported in Gov.

Reply:
Spoke with @gabog, leaving code in.